src/dtls.c

```diff
@@ -290,60 +396,534 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
 #endif
     ret = TlsSessionCacheGetAndLock(sessionID.elements, &sess, &sessRow, 1);
     if (ret == 0 && sess != NULL) {
-        *isValid = 1;
+#if defined(SESSION_CERTS) || (defined(WOLFSSL_TLS13) && \
+                               defined(HAVE_SESSION_TICKET))
+        /* This logic is only for TLS <= 1.2 tickets. Don't accept
+         * TLS 1.3. */
+        if (!IsAtLeastTLSv1_3(sess->version))
+#endif
+        {
+            *resume = 1;
+        }
         TlsSessionCacheUnlockRow(sessRow);
     }
 
     return 0;
 }
 
-static int TlsResumptionIsValid(WOLFSSL* ssl, WolfSSL_CH* ch, byte* isValid)
+static int TlsResumptionIsValid(const WOLFSSL* ssl, WolfSSL_CH* ch,
+                                int* resume)
 {
     int ret;
 
-    *isValid = 0;
 #ifdef HAVE_SESSION_TICKET
-    ret = TlsTicketIsValid(ssl, ch->extension, isValid);
+    ret = TlsTicketIsValid(ssl, ch->extension, resume);
     if (ret != 0)
         return ret;
-    if (*isValid)
+    if (*resume)
         return 0;
 #endif /* HAVE_SESSION_TICKET */
-    ret = TlsSessionIdIsValid(ssl, ch->sessionId, isValid);
+    ret = TlsSessionIdIsValid(ssl, ch->sessionId, resume);
+    return ret;
+}
+#endif /* WOLFSSL_DTLS13 || WOLFSSL_DTLS_NO_HVR_ON_RESUME */
+
+#ifdef WOLFSSL_DTLS13
+static int TlsCheckSupportedVersion(const WOLFSSL* ssl,
+        WolfSSL_CH* ch, byte *isTls13)
+{
+    WolfSSL_ConstVector tlsxSupportedVersions;
+    int ret;
+    ProtocolVersion pv = ssl->version;
+    int tlsxFound;
+
+    ret = TlsxFindByType(&tlsxSupportedVersions, TLSX_SUPPORTED_VERSIONS,
+                         ch->extension, &tlsxFound);
+    if (ret != 0)
+        return ret;
+    if (!tlsxFound) {
+        *isTls13 = 0;
+        return 0;
+    }
+    ret = TLSX_SupportedVersions_Parse(ssl, tlsxSupportedVersions.elements,
+            (word16)tlsxSupportedVersions.size, client_hello, &pv, NULL, NULL);
+    if (ret != 0)
+        return ret;
+    if (IsAtLeastTLSv1_3(pv))
+        *isTls13 = 1;
+    else
+        *isTls13 = 0;
+
+    return 0;
+}
+
+static int CopySupportedGroup(TLSX* src, TLSX** dst, void* heap)
```

Comment:
Belongs in tls.c.
TLSX_SupportedGroup_Copy()

Reply:
Moving.